casual/casual-http/src/test/groovy/se/laz/casual/http/resources/CasualServiceTest.groovy

```diff
@@ -0,0 +1,359 @@
+import se.laz.casual.api.buffer.type.ServiceBuffer
+import se.laz.casual.api.buffer.type.fielded.FieldedTypeBuffer
+import se.laz.casual.api.flags.AtmiFlags
+import se.laz.casual.api.flags.ErrorState
+import se.laz.casual.api.flags.Flag
+import se.laz.casual.api.flags.ServiceReturnState
+import se.laz.casual.api.xa.XID
+import se.laz.casual.http.resources.handlers.CasualServiceCallWorkCreator
+import se.laz.casual.http.resources.handlers.ExceptionHandler
+import se.laz.casual.http.resources.handlers.ExceptionHandlerImpl
+import se.laz.casual.http.resources.handlers.LocalRequestHandler
+import se.laz.casual.http.resources.handlers.RemoteRequestHandler
+import se.laz.casual.jca.inflow.work.CasualServiceCallWork
+import se.laz.casual.network.protocol.messages.CasualNWMessageImpl
+import se.laz.casual.network.protocol.messages.service.CasualServiceCallReplyMessage
+import se.laz.casual.network.protocol.messages.service.CasualServiceCallRequestMessage
+import spock.lang.Shared
+import spock.lang.Specification
+import spock.lang.Unroll
+
+import java.nio.charset.StandardCharsets
+import java.util.concurrent.CompletableFuture
+import java.util.stream.Collectors
+
+class CasualServiceTest extends Specification
+{
+   @Shared
+   def root = '/casual'
+   @Shared
+   def serviceName = 'echo'
+   @Shared
+   def content = '{"msg":"bazinga!"}'
+   @Shared
+   def key = 'FLD_STRING1'
+   @Shared
+   def serviceRegistryLookupServiceDoesNotExist = Mock(ServiceRegistryLookup) {
+      serviceExists(_) >> false
+   }
+   @Shared
+   def serviceRegistryLookupServiceExists = Mock(ServiceRegistryLookup) {
+      serviceExists(_) >> true
+   }
+
+   @Unroll
+   def 'remote service call ok #mimeType #expectedReturnMimeType'()
+   {
+      given:
+      Dispatcher dispatcher = MockDispatcherFactory.createDispatcher()
+      ServiceCaller serviceCaller = Mock(ServiceCaller){
+         1 * makeServiceCall(_, serviceName, Flag.of(AtmiFlags.TPNOTRAN)) >> { CasualBuffer msg, String serviceName, Flag<AtmiFlags> flags ->
+            new ServiceCallResponse(ServiceReturnState.TPSUCCESS, ErrorState.OK, replyBuffer)
+         }
+      }
+      CasualService casualService = new CasualService(serviceCaller, new RemoteRequestHandler(), Mock(LocalRequestHandler), Mock(ExceptionHandler), serviceRegistryLookupServiceDoesNotExist)
+      dispatcher.getRegistry().addSingletonResource(casualService)
+      MockHttpRequest request = MockHttpRequest.post("${root}/${serviceName}")
+              .contentType(mimeType)
+              .content(requestBuffer.getBytes().first)
+      MockHttpResponse response = new MockHttpResponse()
+      expect:
+      dispatcher.invoke(request, response)
+      response.getStatus() == Response.Status.OK.statusCode
+      response.outputHeaders[HttpHeaders.CONTENT_TYPE].first.toString().contains(expectedReturnMimeType)
+      CasualBuffer responseBuffer = creatorFunction(response.output)
+      responseBuffer.getBytes() == requestBuffer.getBytes()
```

Comment:
no check that the local service implementation was not called during a remote call.